### PR TITLE
build: Add `mz-build-tools` crate

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR introduces a new `mz-build-tools` crate which is designed to expose any external tools required in build scripts. Currently the only such tool is `protoc`.

### Motivation

This is a lead up to building Materialize with Bazel. Today we depend on `protobuf-src` which builds `protoc` for us in its build script, and then returns the path of the compiled binary. For Bazel though, we don't want to depend on `protobuf-src` and instead let Bazel build `protoc`.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -83,6 +84,8 @@ members = [
     "src/sql-parser",
     "src/sql-pretty",
     "src/sqllogictest",
+    "src/stash",
+    "src/stash-types",
```
Comment (jkosh44) [inline Cargo.toml:88]: Maybe you need to rebase? These shouldn't exist anymore.

File: src/build-tools/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "mz-build-tools"
+description = "Provides access to tools required in build scripts."
+version = "0.0.0"
+edition.workspace = true
+rust-version.workspace = true
+publish = false
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (jkosh44) [inline src/build-tools/Cargo.toml:9]: Do we usually delete this comment or leave it?

**Review by jkosh44 — APPROVED** [review]

File: src/build-tools/Cargo.toml (hunk 2 above)
Comment (ParkMyCar) [reply src/build-tools/Cargo.toml]: Good call, I removed it

**Review by def- — APPROVED** [review]

File: src/build-tools/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+/// * Bazel runfiles, if we're building with Bazel.
+/// * Bootstraps `protoc` via protobuf-src, if default features are enabled.
+/// * `PROTOC` environment variable, if it's set.
+/// * The system's `$PATH`, via [`which`].
+///
+/// If `protoc` can't be found then this function will panic.
+pub fn protoc() -> PathBuf {
+    cfg_if! {
+        if #[cfg(bazel)] {
+            let r = runfiles::Runfiles::create().unwrap();
```
Comment (guswynn) [inline src/build-tools/src/lib.rs:40]: is this a bazel-provided thing?

**Review by guswynn — APPROVED** [review]
So I understand: when we hook up bazel, the protoc cpp target will be a dep of this crate, and bazel will place the runfiles pointing to the protoc binary in the `runfiles.rs` for us, right?

File: src/build-tools/src/lib.rs (hunk 3 above)
Comment (ParkMyCar) [reply src/build-tools/src/lib.rs:40]: Yeah this is provided by the [Bazel rules_rust](https://github.com/bazelbuild/rules_rust/tree/main/tools/runfiles) project

**Comment by ParkMyCar** [comment]
> So I understand: when we hook up bazel, the protoc cpp target will be a dep of this crate, and bazel will place the runfiles pointing to the protoc binary in the `runfiles.rs` for us, right?

> when we hook up bazel, the protoc cpp target will be a dep of this crate

Exactly!

> bazel will place the runfiles pointing to the protoc binary in the `runfiles.rs` for us, right?

Almost, Bazel will place the compiled `protoc` binary in its runfiles directory, and `runfiles.rs` is how we locate that directory and get a path to `protoc`. [Runfiles](https://bazel.build/extending/rules#runfiles) are Bazel's way of providing files at runtime, as opposed to compile time. Due to Bazel's sandboxing the absolute location of the runfiles can change, which is why we need to use `runfiles.rs` to look up its location

**Review by benesch — COMMENTED** [review]
I'm late to the party here, but this looks great! Thanks for working on this incrementally.

File: src/catalog/Cargo.toml [hunk 4]
```diff
@@ -73,10 +73,13 @@ tokio-postgres = { version = "0.7.8" }
 [build-dependencies]
 anyhow = "1.0.66"
 md-5 = "0.10.5"
-protobuf-src = "1.1.0"
```
Comment (antiguru) [inline src/catalog/Cargo.toml:76]: Late to comment, but we should probably add a check to [ci/test/lint-deps.sh](https://github.com/MaterializeInc/materialize/blob/5aa09404319513a4a9ad913849c396af08907bc2/ci/test/lint-deps.sh) to make sure nobody introduces a dependency on `protobuf-src` again.

Comment (antiguru) [reply src/catalog/Cargo.toml:76]: Here's a diff
```diff
diff --git a/ci/test/lint-deps.sh b/ci/test/lint-deps.sh
index 85064519b4..365adbe01a 100755
--- a/ci/test/lint-deps.sh
+++ b/ci/test/lint-deps.sh
@@ -37,6 +37,7 @@ crates=(
     # be disabled using --no-default-features
     tikv_jemalloc_ctl
     tikv_jemallocator
+    protobuf_src
 )

 rewrite=false
(
```

Comment (ParkMyCar) [reply src/catalog/Cargo.toml:76]: Great idea! Just put up https://github.com/MaterializeInc/materialize/pull/26666 to do this

**Pull request merged**